### DISCUSSION TITLE
Using Running sum{} and last{} together in QQL

### DISCUSSION THREAD

**Discussion by ShivaniSK** [finos/TimeBase-CE#245] [Q&A]
HI @alex-karpovich ,
I am trying to achieve a result where i want the data to be rolled up over 1m where I need the last record of each minute in response. Along with it, I also want to calculate a cumulative sum of a column at every minute.
I am trying below query but it does not help

select running sum{} (ask_size) as cumulative_sum, last{} (ask_size) as size, last{} (ask_price) as price, level from order_book_depth RESET over TIME (1m) where timestamp BETWEEN 1661513400000 AND 1661514600000 AND SYMBOL IN ('EURUSD') 

**Expected output:**
  timestamp           symbol  size   price   cumulative_sum
0,2022-08-26 11:31:00,EURUSD,1000000,1.00076,1,1000000
1,2022-08-26 11:31:00,EURUSD,1000000,1.00073,2,2000000
2,2022-08-26 11:31:00,EURUSD,1000000,1.00068,3,3000000
3,2022-08-26 11:31:00,EURUSD,1000000,1.00063,4,4000000
4,2022-08-26 11:31:00,EURUSD,1000000,1.00058,5,5000000
5,2022-08-26 11:32:00,EURUSD,5000000,1.001,1,1000000
6,2022-08-26 11:32:00,EURUSD,5000000,1.00095,2,2000000
7,2022-08-26 11:32:00,EURUSD,7000000,1.0009,3,3000000
8,2022-08-26 11:32:00,EURUSD,5000000,1.00085,4,4000000
9,2022-08-26 11:32:00,EURUSD,4000000,1.0008,5,5000000

**Actual output**
>_,TIMESTAMP,SYMBOL,TYPE,CUMULATIVE_SUM,SIZE,PRICE,level
0,2022-08-26 11:30:00.010,EURUSD,6000000,6000000,1.00075,1
1,2022-08-26 11:30:00.010,EURUSD,12000000,6000000,1.0008,2
2,2022-08-26 11:30:00.010,EURUSD,16000000,4000000,1.00085,3
3,2022-08-26 11:30:00.010,EURUSD,20000000,4000000,1.0009,4
4,2022-08-26 11:30:00.010,EURUSD,25000000,5000000,1.00095,5
5,2022-08-26 11:30:00.015,EURUSD,31000000,6000000,1.00075,1
6,2022-08-26 11:30:00.015,EURUSD,38000000,7000000,1.0008,2
7,2022-08-26 11:30:00.015,EURUSD,42000000,4000000,1.00085,3
8,2022-08-26 11:30:00.015,EURUSD,46000000,4000000,1.0009,4
9,2022-08-26 11:30:00.015,EURUSD,51000000,5000000,1.00095,5

Can you please help me out.

Regards,
Shivani

**Comment by alex-karpovich** [comment]
@ShivaniSK  Please share source data sample

**Reply by ShivaniSK** [reply]
Hi @alex-karpovich ,
Please find below source data sample

==>  select ask_size, ask_price, level from order_book  where timestamp BETWEEN 1661513400000 AND 1661514600000 AND SYMBOL IN ('ABC')
>_,TIMESTAMP,SYMBOL,TYPE,ask_size,ask_price,level
0,2022-08-26 11:30:00.010,ABC,6000000,1.00075,1
1,2022-08-26 11:30:00.010,ABC,6000000,1.0008,2
2,2022-08-26 11:30:00.010,ABC,4000000,1.00085,3
3,2022-08-26 11:30:00.010,ABC,4000000,1.0009,4
4,2022-08-26 11:30:00.010,ABC,5000000,1.00095,5
5,2022-08-26 11:30:00.015,ABC,6000000,1.00075,1
6,2022-08-26 11:30:00.015,ABC,7000000,1.0008,2
7,2022-08-26 11:30:00.015,ABC,4000000,1.00085,3
8,2022-08-26 11:30:00.015,ABC,4000000,1.0009,4
9,2022-08-26 11:30:00.015,ABC,5000000,1.00095,5

**Comment by alex-karpovich** [comment]
Greetings @ShivaniSK 

`
select sum{} (ask_size) as cumulative_sum, last{} (ask_size) as size, last{} (ask_price) as price, level from order_book_depth RESET over TIME (1m) where timestamp BETWEEN 1661513400000 AND 1661514600000 AND SYMBOL IN ('EURUSD')
`

do not use 'running' - it used to calculate result for each record

https://kb.timebase.info/community/development/qql/QQL%205.5/qql-tut-functions#stateful-functions

**Reply by ShivaniSK** [reply]
Hello @alex-karpovich ,
The above query sugessted by you is not working as expected. Its giving the total sum of the ask_size column, instead of doing cumulative sum.

**Actual output:**

==> select sum{} (ask_size) as cumulative_sum, last{} (ask_size) as size, last{} (ask_price) as price, level from order_book_depth RESET over TIME (1m) where timestamp BETWEEN 1661513400000 AND 1661514600000 AND SYMBOL IN ('EURUSD')
>_,TIMESTAMP,SYMBOL,TYPE,CUMULATIVE_SUM,SIZE,PRICE,level
0,2022-08-26 11:31:00,EURUSD,7166000000,1000000,1.00076,5
1,2022-08-26 11:32:00,EURUSD,8093000000,5000000,1.001,5
2,2022-08-26 11:33:00,EURUSD,7388000000,6000000,1.0009,5
3,2022-08-26 11:34:00,EURUSD,15647000000,1000000,1.00049,5
4,2022-08-26 11:35:00,EURUSD,12243000000,1000000,1.00036,5
5,2022-08-26 11:36:00,EURUSD,9021000000,4000000,1.0003,5
6,2022-08-26 11:37:00,EURUSD,8971000000,2000000,1.00035,5
7,2022-08-26 11:38:00,EURUSD,8953000000,1000000,0.99999,5
8,2022-08-26 11:39:00,EURUSD,11665000000,1000000,0.99998,5
9,2022-08-26 11:40:00,EURUSD,9949000000,3000000,0.9998,5

**Expected output**

timestamp symbol size price cumulative_sum
0,2022-08-26 11:31:00,EURUSD,1000000,1.00076,1,1000000
1,2022-08-26 11:31:00,EURUSD,1000000,1.00073,2,2000000
2,2022-08-26 11:31:00,EURUSD,1000000,1.00068,3,3000000
3,2022-08-26 11:31:00,EURUSD,1000000,1.00063,4,4000000
4,2022-08-26 11:31:00,EURUSD,1000000,1.00058,5,5000000
5,2022-08-26 11:32:00,EURUSD,5000000,1.001,1,1000000
6,2022-08-26 11:32:00,EURUSD,5000000,1.00095,2,2000000
7,2022-08-26 11:32:00,EURUSD,7000000,1.0009,3,3000000
8,2022-08-26 11:32:00,EURUSD,5000000,1.00085,4,4000000
9,2022-08-26 11:32:00,EURUSD,4000000,1.0008,5,5000000

**Reply by alex-karpovich** [reply]
@ShivaniSK 

1. in your expected output is unclear how you calculate cumulative sum for this rows - please explain

5,2022-08-26 11:32:00,EURUSD,5000000,1.001,1,**1000000** (should be 5,000,000?)
6,2022-08-26 11:32:00,EURUSD,5000000,1.00095,2,**2000000** (should be 10,000,000?)
7,2022-08-26 11:32:00,EURUSD,7000000,1.0009,3,**3000000** (should be 17,000,000?)
8,2022-08-26 11:32:00,EURUSD,5000000,1.00085,4,**4000000** (?)
9,2022-08-26 11:32:00,EURUSD,4000000,1.0008,5,**5000000** (?)


2. to be more precise could you please give me EXACT sample of your data  (say 20-30 rows) that we can run same queries to verify output

3.  explaining query `select running sum{} (ask_size) as cumulative_sum ... reset over time(1m)`
     in this construction sum is `sliding window` function with fixed range 1 minute - it will calculates cumulative sum for the each message belongs to the [T, T+1) time interval
